docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,403 @@
+}
+```
+
+#### Pruning Service
+
+This gRPC service is the only novel service proposed in this ADR, and
+effectively gives a single external caller (e.g. a data companion) a say in how
+the node prunes its data.
+
+```protobuf
+syntax = "proto3";
+
+package tendermint.services.pruning.v1;
+
+// PruningService provides privileged access to specialized pruning
+// functionality on the CometBFT node to help control node storage.
+service PruningService {
+    // SetBlockRetainHeightRequest indicates to the node that it can safely
+    // prune all block data up to the specified retain height.
+    //
+    // The lower of this retain height and that set by the application in its
+    // Commit response will be used by the node to determine which heights' data
+    // can be pruned.
+    rpc SetBlockRetainHeight(SetBlockRetainHeightRequest) returns (SetBlockRetainHeightResponse);
+
+    // GetBlockRetainHeight returns information about the retain height
+    // parameters used by the node to influence block retention/pruning.
+    rpc GetBlockRetainHeight(GetBlockRetainHeightRequest) returns (GetBlockRetainHeightResponse);
+
+    // SetBlockResultsRetainHeightRequest indicates to the node that it can
+    // safely prune all block results data up to the specified height.
+    //
+    // The node will always store the block results for the latest height to
+    // help facilitate crash recovery.
+    rpc SetBlockResultsRetainHeight(SetBlockResultsRetainHeightRequest) returns (SetBlockResultsRetainHeightResponse);
+
+    // GetBlockResultsRetainHeight returns information about the retain height
+    // parameters used by the node to influence block results retention/pruning.
+    rpc GetBlockResultsRetainHeight(GetBlockResultsRetainHeightRequest) returns (GetBlockResultsRetainHeightResponse);
+}
+
+message SetBlockRetainHeightRequest {
+    uint64 height = 1;
+}
+
+message SetBlockRetainHeightResponse {}
+
+message GetBlockRetainHeightRequest {}
+
+message GetBlockRetainHeightResponse {
+    // The retain height set by the application.
+    uint64 app_retain_height = 1;
+
+    // The retain height set via the pruning service (e.g. by the data
+    // companion) specifically for blocks.
+    uint64 pruning_service_retain_height = 2;
+}
+
+message SetBlockResultsRetainHeightRequest {
+    uint64 height = 1;
+}
+
+message SetBlockResultsRetainHeightResponse {}
+
+message GetBlockResultsRetainHeightRequest {}
+
+message GetBlockResultsRetainHeightResponse {
+    // The retain height set by the pruning service (e.g. by the data
+    // companion) specifically for block results.
+    uint64 pruning_service_retain_height = 1;
+}
+```
+
+With this API design, it is technically possible for an integrator to attach
+multiple data companions to the node, but only one of their retain heights will
+be considered by the node.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
```

Comment:
Those additions are to CometBFT's configuration file, right?

Reply:
Yes, these are sections added to `config.toml` . I can make it more specific in the text if you prefer.